File: 1034-subarrays-with-k-different-integers/subarrays-with-k-different-integers.cpp
class Solution {
public:

    int subarraysWithKAtmost(vector<int>& nums,int k){
        unordered_map<int,int> mp;
        int j=0,res=0;
        for(int i=0;i<nums.size();i++){
            if(mp[nums[i]]==0){
                k--;
            }
            mp[nums[i]]++;

            while(k<0){
                mp[nums[j]]--;
                if(mp[nums[j]]==0){k++;}
                j++;
            }
            res += i-j+1;
        }
        return res;
    }

    int subarraysWithKDistinct(vector<int>& nums, int k) {
        
        return subarraysWithKAtmost(nums,k) - subarraysWithKAtmost(nums,k-1);
    }
};

File: 1034-subarrays-with-k-different-integers/subarrays-with-k-different-integers.java
class Solution {
    public int findArraysWithAtmostKDistinct(int[] nums,int k){
        HashMap<Integer,Integer> mp= new HashMap<>();
        int ans = 0;
        int i=0,j=0;
        for(i=0,j=0;i<nums.length;i++){
            mp.put(nums[i],mp.containsKey(nums[i]) ? mp.get(nums[i])+1 : 1);
            while(mp.size()>k){
                mp.put(nums[j],mp.get(nums[j])-1);
                if(mp.get(nums[j])==0){
                    mp.remove(nums[j]);
                }
                j++;
            }
            ans += i-j+1;
        }
        return ans;
    }

    public int subarraysWithKDistinct(int[] nums, int k) {
        return findArraysWithAtmostKDistinct(nums,k) - findArraysWithAtmostKDistinct(nums,k-1);
    }
}